Apps/Common/src/Services/EmailQueueService.cs

```diff
@@ -0,0 +1,120 @@
+        {
+            this.emailDelegate = emailDelegate;
+        }
+
+        /// <inheritdoc />
+        public void QueueEmail(string toEmail, string templateName, Dictionary<string, string> keyValues)
+        {
+            this.QueueEmail(toEmail, this.GetEmailTemplate(templateName), keyValues);
+        }
+
+        /// <inheritdoc />
+        public void QueueEmail(string toEmail, EmailTemplate emailTemplate, Dictionary<string, string> keyValues)
+        {
+            this.QueueEmail(this.ProcessTemplate(toEmail, emailTemplate, keyValues));
+        }
+
+        /// <inheritdoc />
+        public void QueueEmail(Email email)
+        {
+            this.emailDelegate.InsertEmail(email);
+            BackgroundJob.Enqueue<IEmailJob>(j => j.SendEmail(email.Id));
+        }
+
+        /// <inheritdoc />
+        public void QueueInviteEmail(string hdid, string toEmail, string templateName, Dictionary<string, string> keyValues)
+        {
+            this.QueueInviteEmail(hdid, toEmail, this.GetEmailTemplate(templateName), keyValues);
+        }
+
+        /// <inheritdoc />
+        public void QueueInviteEmail(string hdid, string toEmail, EmailTemplate emailTemplate, Dictionary<string, string> keyValues)
+        {
+            EmailInvite invite = new EmailInvite();
+            invite.HdId = hdid;
+            invite.Email = this.ProcessTemplate(toEmail, emailTemplate, keyValues);
+            this.QueueInviteEmail(invite);
+        }
+
+        /// <inheritdoc />
+        public void QueueInviteEmail(EmailInvite invite)
+        {
+            this.emailDelegate.InsertEmailInvite(invite);
+            BackgroundJob.Enqueue<IEmailJob>(j => j.SendEmail(invite.Email.Id));
+        }
+
+        /// <inheritdoc />
+        public EmailTemplate GetEmailTemplate(string templateName)
+        {
+            return this.emailDelegate.GetEmailTemplate(templateName);
+        }
+
+        /// <inheritdoc />
+        public Email ProcessTemplate(string toEmail, EmailTemplate emailTemplate, Dictionary<string, string> keyValues)
+        {
+            Email email = ParseTemplate(emailTemplate, keyValues);
+            email.To = toEmail;
+            return email;
+        }
+
+        private static Email ParseTemplate(EmailTemplate emailTemplate, Dictionary<string, string> keyValues)
+        {
+            Contract.Requires(emailTemplate != null);
+            Email email = new Email();
+            email.From = emailTemplate.From;
+            email.Priority = emailTemplate.Priority;
+            email.Subject = ProcessTemplateString(emailTemplate.Subject, keyValues);
+            email.Body = ProcessTemplateString(emailTemplate.Body, keyValues);
+            email.FormatCode = emailTemplate.FormatCode;
+            return email;
+        }
+
+        private static string ProcessTemplateString(string template, Dictionary<string, string> data)
+        {
+            return Regex.Replace(template, "\\$\\{(.*?)\\}", m =>
```

Comment:
this looks quite confusing, either lets break it up or add documentation about it is doing